src/configuration/configuration.ts

```diff
@@ -0,0 +1,280 @@
+    this._value = value;
+    this.source = OptionValueSource.Runtime;
+  }
+
+  public source: OptionValueSource = OptionValueSource.Default  ;
+  static globalOptionUpdateHandler: () => void;
+}
+
+export class OptionMap {
+  public static allOptions: { [key: string]: BaseOptionDetails } = {};
+}
+
+export function RegisterOption(optionDetails: typeof BaseOptionDetails): void {
+  let newOption = new optionDetails();
+  newOption.initialize();
+  OptionMap.allOptions[newOption.id] = newOption;
+}
+
+@RegisterOption
+class OptionUseSolidBlockCursor extends BaseOptionDetails {
+  id = "useSolidBlockCursor";
+  defaultValue = false;
+}
+
+@RegisterOption
+class OptionUseControlKeys extends BaseOptionDetails {
+  id = "useCtrlKeys";
+  defaultValue = false;
+}
+
+@RegisterOption
+class OptionScroll extends BaseOptionDetails {
+  id = "scroll";
+  defaultValue = 20;
+}
+
+@RegisterOption
+class OptionHightlightSearch extends BaseOptionDetails {
+  id = "hlsearch";
+  defaultValue = false;
+}
+
+@RegisterOption
+class OptionIgnoreCase extends BaseOptionDetails {
+  id = "ignorecase";
+  defaultValue = true;
+}
+
+@RegisterOption
+class OptionSmartCase extends BaseOptionDetails {
+  id = "smartcase";
+  defaultValue = true;
+}
+
+@RegisterOption
+class OptionTabStop extends BaseOptionDetails {
+  id = "tabstop"; // aka, tabSize
+  defaultValue = 8; // Vim default value but here we always use Code configuration
+
+  fetchCodeConfiguration() {
+    const vimOptionInCode = vscode.workspace.getConfiguration("vim").get<number>(this.id, undefined);
+
+    return vimOptionInCode !== undefined ? vimOptionInCode : this.getTabSizeFromCode();
+  }
+
+  initialize() {
+    super.initialize();
+    // TODO we should update through API instead of tweaking editor's copy of config.
+    const options = vscode.window.activeTextEditor.options;
+    options.tabSize = this.value as number;
+    vscode.window.activeTextEditor.options = options;
+  }
+
+  public get value() {
+    if (this.source === OptionValueSource.Runtime) {
+      return this._value;
+    } else {
+      return this.getTabSizeFromCode();
+    }
+  }
+
+  public set value(value: OptionValue) {
+    super.value = value;
+
+    if (!vscode.window.activeTextEditor) {
+      // TODO: We should set configuration by API, which is not currently supported.
+      return;
+    }
+
+    const oldOptions = vscode.window.activeTextEditor.options;
+    oldOptions.tabSize = value as number;
+    vscode.window.activeTextEditor.options = oldOptions;
+  }
+
+  private getTabSizeFromCode(): number {
+    if (vscode.window.activeTextEditor) {
+      return vscode.window.activeTextEditor.options.tabSize! as number;
+    } else {
+      return vscode.workspace.getConfiguration("editor").get("tabSize", this.defaultValue);
+    }
+  }
+}
+
+@RegisterOption
+class OptionInsertSpaces extends BaseOptionDetails {
+  id = "expandtab"; // aka insertSpaces
+  defaultValue = false;
+
+  fetchCodeConfiguration() {
```

Comment:
There is a bunch of duplication between this and OptionTabStop that you're going to need to remove as well.